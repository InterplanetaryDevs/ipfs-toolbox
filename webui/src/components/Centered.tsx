import React, {PropsWithChildren} from 'react';
import {Box} from '@mui/material';

export function Centered(props: PropsWithChildren<{}>) {
	return <Box sx={{
		display: 'flex',
		alignItems: 'center',
		justifyContent: 'space-around',
		width: '100%',
		height: '100%',
	}}>
		{props.children}
	</Box>;
}
