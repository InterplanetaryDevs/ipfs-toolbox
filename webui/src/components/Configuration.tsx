import {ITool} from '../App';
import React from 'react';

export const ConfigurationDefinition: ITool = {
	tool: <Configuration/>,
	name: 'Configuration'
};

export function Configuration() {
	return <div></div>;
}