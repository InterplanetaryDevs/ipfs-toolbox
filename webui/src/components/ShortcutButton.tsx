import React from 'react';
import BoltIcon from '@mui/icons-material/Bolt';
import {Button, IconButton} from '@mui/material';
import {IShortCut} from '../services/ShortcutService/IShortCut';

export function ShortcutButton(props: { shortcut: IShortCut, iconOnly?: boolean }) {
	function onClick() {
		props.shortcut.action();
	}

	if (props.iconOnly) {
		return <IconButton onClick={onClick}>{props.shortcut.icon ?? <BoltIcon/>}</IconButton>;
	}
	return <Button onClick={onClick}>
		{props.shortcut.icon}
		{props.shortcut.name}
	</Button>;
}
