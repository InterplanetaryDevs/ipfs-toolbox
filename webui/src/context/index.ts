export {useIpfsCluster, IpfsClusterContextProvider} from './IpfsClusterContext';
export type {IIpfsClusterContext} from './IpfsClusterContext';
export {useIpfs, IpfsContextProvider} from './IpfsContext';
export type {IIpfsContext} from './IpfsContext';
export {useToolBox, ToolBoxContextProvider} from './ToolBoxContext';
export type {IToolBoxContext} from './ToolBoxContext';