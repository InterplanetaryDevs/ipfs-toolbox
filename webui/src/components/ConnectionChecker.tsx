import {Alert, Box, Button, Card, CardActions, CardContent, CardHeader} from '@mui/material';
import React, {PropsWithChildren} from 'react';
import {Centered} from './Centered';
import ReplayIcon from '@mui/icons-material/Replay';
import {IConnectionCheckAble} from '../services/IConnectionCheckAble';

type ConnectionCheckerProps = PropsWithChildren<{
	context: IConnectionCheckAble
}>

export function ConnectionChecker({context, children}: ConnectionCheckerProps) {
	if (!context.connected) {
		return <Centered>
        <Card sx={{width: '50%'}}>
            <CardHeader title={'Connection Failed!'}/>
            <CardContent>
							{context.notConnectedMessage ?? <Alert severity="error">Failed to connect!</Alert>}
            </CardContent>
            <CardActions>
								{context.checking && <Alert severity={'warning'}>Retrying...</Alert>}
                <Box sx={{flexGrow: 1}}/>
                <Button startIcon={<ReplayIcon/>}>Retry</Button>
            </CardActions>
        </Card>
    </Centered>;
	}

	return <div>{children}</div>;
}
