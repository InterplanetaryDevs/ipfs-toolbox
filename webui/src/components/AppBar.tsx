import {AppBar as MuiAppBar, Chip, IconButton, Popover, Stack, Toolbar, Typography} from '@mui/material';
import MenuIcon from '@mui/icons-material/Menu';
import {useIpfs} from '../context/IpfsContext';
import {useToolBox} from '../context/ToolBoxContext';
import {ConnectionTextField} from './ConnectionTextField';
import {useIpfsCluster} from '../context/IpfsClusterContext';
import React, {PropsWithoutRef} from 'react';
import InfoIcon from '@mui/icons-material/Info';
import {IKeyBind} from '../services/ShortcutService';

interface IAppBarProps {
}


function ShortcutDisplay({keyBind}: PropsWithoutRef<{ keyBind: IKeyBind }>) {
	return <div>
		{keyBind.ctrl && <><Chip label={'CTRL'}/> + </>}
		{keyBind.alt && <><Chip label={'ALT'}/> + </>}
		{keyBind.shift && <><Chip label={'SHIFT'}/> + </>}
		<Chip label={keyBind.key.replace(' ', 'space').toUpperCase()}/>
	</div>;
}

export function AppBar(props: IAppBarProps) {
	const {tool, menu, setMenuOpen, shortcutService} = useToolBox();
	const {connected: ipfsConnected, apiUrl, setApiUrl, checking: ipfsChecking} = useIpfs();
	const {
		apiUrl: clusterApiUrl,
		setApiUrl: clusterSetApiUrl,
		connected: clusterConnected,
		checking: clusterChecking
	} = useIpfsCluster();

	const [anchorEl, setAnchorEl] = React.useState<HTMLButtonElement | null>(null);
	const open = Boolean(anchorEl);

	const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
		setAnchorEl(event.currentTarget);
	};

	const handleClose = () => {
		setAnchorEl(null);
	};


	return <MuiAppBar position="fixed">
		<Toolbar>
			<IconButton
				color="inherit"
				aria-label="open drawer"
				onClick={() => setMenuOpen(true)}
				edge="start"
			>
				<MenuIcon/>
			</IconButton>
			<Typography variant="h6" noWrap component="div">
				{tool.name}
			</Typography>
			{menu}
			<div style={{flexGrow: 1}}/>
			<Stack spacing={2} direction={'row'}>
				<ConnectionTextField
					value={apiUrl}
					onChange={setApiUrl}
					label={'IPFS Url'}
					placeholder={'/ip4/127.0.0.1/tcp/5001'}
					connected={ipfsConnected}
					checking={ipfsChecking}
				/>
				<ConnectionTextField
					value={clusterApiUrl}
					onChange={clusterSetApiUrl}
					label={'IPFS Cluster Url'}
					placeholder={'http://localhost:9094'}
					connected={clusterConnected}
					checking={clusterChecking}
				/>
				<IconButton onClick={handleClick}>
					<InfoIcon/>
				</IconButton>
				<Popover
					open={open}
					anchorEl={anchorEl}
					onClose={handleClose}
					anchorOrigin={{
						vertical: 'bottom',
						horizontal: 'right',
					}}
				>
					<Stack>
						{shortcutService.getShortcuts().map((v, i) => <div style={{display: 'flex'}}>
							<div style={{flexGrow: 1}}>{v.name}</div>
							<ShortcutDisplay keyBind={v.keyBind}/></div>)}
					</Stack>
				</Popover>
			</Stack>
		</Toolbar>
	</MuiAppBar>;
}
