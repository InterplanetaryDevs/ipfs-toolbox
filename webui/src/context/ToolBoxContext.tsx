import {createContext, PropsWithChildren, useCallback, useContext, useEffect, useState} from 'react';
import {ITool} from '../App';
import {DashboardDefinition} from '../components/Dashboard';
import {ConfigurationDefinition} from '../components/Configuration';

export interface IToolBoxContext {
	tool: ITool,
	setTool: (value: (((prevState: ITool) => ITool) | ITool)) => void,
	isMenuOpen: boolean,
	setMenuOpen: (value: (((prevState: boolean) => boolean) | boolean)) => void,
	menu: JSX.Element,
	setMenu: (value: (((prevState: (JSX.Element | undefined)) => (JSX.Element | undefined)) | JSX.Element | undefined)) => void,
}

const ToolBoxContext = createContext<IToolBoxContext>({} as IToolBoxContext);

export function ToolBoxContextProvider(props: PropsWithChildren) {
	const [tool, setTool] = useState(DashboardDefinition);
	const [menu, setMenu] = useState<JSX.Element>();
	const [isMenuOpen, setMenuOpen] = useState(false);

	const handleKeyPress = useCallback((event: KeyboardEvent) => {
		if (event.key == 'h' && event.ctrlKey) {
			event.preventDefault();
			setTool(DashboardDefinition);
		} else if (event.key == ',' && event.ctrlKey) {
			event.preventDefault();
			setTool(ConfigurationDefinition);
		}
		console.log(`Key pressed: ${event.key}`);
	}, []);

	useEffect(() => {
		document.addEventListener('keydown', handleKeyPress);

		return () => {
			document.removeEventListener('keydown', handleKeyPress);
		};
	}, [handleKeyPress]);

	return <ToolBoxContext.Provider value={{
		tool,
		setTool,
		isMenuOpen,
		setMenuOpen,
		menu,
		setMenu,
	}}>
		{props.children}
	</ToolBoxContext.Provider>;
}

export function useToolBox(): IToolBoxContext {
	return useContext<IToolBoxContext>(ToolBoxContext);
}
