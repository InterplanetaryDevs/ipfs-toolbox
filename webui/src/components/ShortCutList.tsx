import {Box, Stack, Typography} from '@mui/material';
import React from 'react';
import {useToolBox} from '../context/ToolBoxContext';
import {IShortCut} from '../services/ShortcutService/IShortCut';
import {ShortCutDisplay} from './ShortCutDisplay';

export function ShortCutList() {
	const {shortcutService} = useToolBox();

	return <Box>
		<Stack spacing={1}>
			{Object.entries(shortcutService.getShortCuts()
				.reduce((group, product) => {
					const category = product.category ?? 'unknown';
					group[category] = group[category] ?? [];
					group[category].push(product);
					return group;
				}, {} as { [key: string]: IShortCut[] }))
				.map(([category, shortcuts]) => <Stack>
					<Typography variant={'caption'}>{category}</Typography>
					{shortcuts.map((v, i) => (<Box
						sx={{display: 'flex', borderBottomWidth: 1, borderBottomColor: 'primary.dark'}}>
						<div style={{flexGrow: 1}}>{v.name}</div>
						<ShortCutDisplay keyBind={v.keyBind}/>
					</Box>))}
				</Stack>)}
		</Stack>
	</Box>;
}
