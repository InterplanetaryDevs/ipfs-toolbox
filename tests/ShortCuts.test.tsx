import {SnackbarProvider} from 'notistack';
import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';
import {MemoryConfigurationStore} from 'testing';
import {ShortcutButton} from 'webui/src/components/ShortcutButton';
import {ShortCutDisplay} from 'webui/src/components/ShortCutDisplay';
import {ShortCutList} from 'webui/src/components/ShortCutList';
import {ToolBoxContextProvider} from 'webui/src/context/ToolBoxContext';

describe('ShortCuts', () => {
	it('Renders List', async () => {
		const component = renderer.create(<MemoryRouter>
			<SnackbarProvider>
				<ToolBoxContextProvider store={new MemoryConfigurationStore()}>
					<ShortCutList/>
				</ToolBoxContextProvider>
			</SnackbarProvider>
		</MemoryRouter>);

		expect(component).toMatchSnapshot();
	});

	it('Renders Display', async () => {
		const component = renderer.create(<ShortCutDisplay keyBind={{
			key: 'h',
			ctrl: true,
			alt: true,
		}}/>);

		expect(component).toMatchSnapshot();
	});

	it('Shortcutbutton works', async () => {
		let called = 0;
		const component = renderer.create(<ShortcutButton shortcut={{
			action: () => {
				called++;
			},
			name: 'Test shortcut',
			keyBind: {
				key: 'h',
				ctrl: true,
				alt: true,
			},
		}}/>);

		expect(called).toBe(0);

		component.root.findByType('button').props.onClick();

		expect(called).toBe(1);

		expect(component).toMatchSnapshot();
	});

	it('Shortcutbutton iconOnly works', async () => {
		let called = 0;
		const component = renderer.create(<ShortcutButton iconOnly={true} shortcut={{
			action: () => {
				called++;
			},
			name: 'Test shortcut',
			keyBind: {
				key: 'h',
				ctrl: true,
				alt: true,
			},
		}}/>);

		expect(called).toBe(0);

		component.root.findByType('button').props.onClick();

		expect(called).toBe(1);

		expect(component).toMatchSnapshot();
	});
});