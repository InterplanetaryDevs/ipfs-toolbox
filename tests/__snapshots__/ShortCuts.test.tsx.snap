// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShortCuts Renders Display 1`] = `
<div>
  <div
    className="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-w66kx-MuiChip-root"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <span
      className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
    >
      CTRL
    </span>
  </div>
   + 
  <div
    className="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-w66kx-MuiChip-root"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <span
      className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
    >
      ALT
    </span>
  </div>
   + 
  <div
    className="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-w66kx-MuiChip-root"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <span
      className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
    >
      H
    </span>
  </div>
</div>
`;

exports[`ShortCuts Renders List 1`] = `
<div
  className="MuiBox-root css-0"
>
  <div
    className="css-1okj3ks-MuiStack-root"
  />
</div>
`;

exports[`ShortCuts Shortcutbutton iconOnly works 1`] = `
<button
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
    data-testid="BoltIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M11 21h-1l1-7H7.5c-.58 0-.57-.32-.38-.66.19-.34.05-.08.07-.12C8.48 10.94 10.42 7.54 13 3h1l-1 7h3.5c.49 0 .56.33.47.51l-.07.15C12.96 17.55 11 21 11 21z"
    />
  </svg>
  <span
    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`ShortCuts Shortcutbutton works 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  Test shortcut
</button>
`;
